### ISSUE TITLE
Kernel.--/2 returns an invalid result

### ISSUE THREAD

**Issue by okoriko** [elixir-lang/elixir#9161]
### Precheck

* Do not use the issue tracker for help or support (try Elixir Forum, Stack Overflow, IRC, etc.)
* For proposing a new feature, please start a discussion on the Elixir Core mailing list: https://groups.google.com/group/elixir-lang-core
* For bugs, do a quick search and make sure the bug has not yet been reported
* Please disclose security vulnerabilities privately at elixir-security@googlegroups.com
* Finally, be nice and have fun!

### Environment

* Elixir & Erlang/OTP versions (elixir --version):

```
Erlang/OTP 21 [erts-10.2.3] [source] [64-bit] [smp:6:6] [ds:6:6:10] [async-threads:1] [hipe]

Interactive Elixir (1.8.0) - press Ctrl+C to exit (type h() ENTER for help)
```

* Operating system: Ubuntu 18.04

### Current behavior
I'm not sure if if it's me not understanding the expected behavior of `--` (as explained in https://hexdocs.pm/elixir/Kernel.html#--/2 ), or if it is a bug in Elixir or Erlang. 

It can be reproduced in `iex` with:

```elixir
ids = [2985095936, 47540628, 135460048, 1266126295, 240535295, 115724671, 161800351,
 4187206564, 4178142725, 234897063, 14773162, 6662515191, 133150693, 378034895,
 1874402262, 3507611978, 22850922, 415521280, 253360400, 71683243]

Enum.count(ids)
# 20

Enum.count(Enum.uniq(ids))
# 20

ids -- ids
# [6662515191]

sorted = ids |> Enum.sort()
sorted -- sorted
# [1874402262, 6662515191]

set = MapSet.new(ids)
MapSet.difference(set, set)
# #MapSet<[]>
```

I couldn't reproduce this behavior with less than 16 elements:
```elixir
Enum.find(1..100_000, fn _ -> sample = Enum.shuffle(ids) |> Enum.take(16); Enum.count(sample -- sample) > 0 end)
# nil

Enum.find(1..100_000, fn _ -> sample = Enum.shuffle(ids) |> Enum.take(17); Enum.count(sample -- sample) > 0 end)
# 1
```

### Expected behavior

In this case, `ids` is a list of unique integer and I would have expected that `--` produces the same result as `MapSet.difference/2`.

**Comment by okoriko** [comment]
Seems like I should maybe report it in erlang: https://github.com/elixir-lang/elixir/blob/v1.9.0/lib/elixir/lib/kernel.ex#L1369

Feel free to close here if so :+1:

**Comment by josevalim** [comment]
@okoriko yes, please report it on bugs.erlang.org. I can also reproduce it on OTP 22. Here is a minimal example:

```erl
Eshell V10.4  (abort with ^G)
1> Ids = [2985095936, 47540628, 135460048, 1266126295, 240535295, 115724671, 161800351,
1>  4187206564, 4178142725, 234897063, 14773162, 6662515191, 133150693, 378034895,
1>  1874402262, 3507611978, 22850922, 415521280, 253360400, 71683243].
[2985095936,47540628,135460048,1266126295,240535295,
 115724671,161800351,4187206564,4178142725,234897063,
 14773162,6662515191,133150693,378034895,1874402262,
 3507611978,22850922,415521280,253360400,71683243]
2> length(Ids).
20
3> length(lists:usort(Ids)).
20
4> Ids -- Ids.
[6662515191]
```

**Comment by chulkilee** [comment]
Reported to https://bugs.erlang.org/browse/ERL-986

**Comment by jhogberg** [comment]
Thanks for your report!

Sorry about the late reply, I'm having trouble logging in to bugs.erlang.org after [it was migrated to a new server](https://bugs.erlang.org/browse/ERL-987) a few days back. I had hoped it would be solved by now, but since it's not I figured it's best to reply here instead.

When a large list is passed to the right-hand side of `--/2`, it's converted into a red-black tree to make lookups faster. The bug was caused by passing a subtly wrong comparison function to the tree; it expected something that returned an `int` but got something returning an `Sint` (`erts_cmp`), which could overflow and make the tree inconsistent.

Patches are on the way for 21 and 22.
